jax/experimental/shard_map.py

```diff
@@ -0,0 +1,912 @@
+
+def _unmatch(mesh, src_tup, x):
+  src = _names_to_pspec(dict(src_tup))
+  dst = P(mesh.axis_names)
+  return shard_map(_add_singleton, mesh, (src,), dst)(x)
+
+def _check_names(names: Sequence[AxisNames], avals: Sequence[core.ShapedArray]
+                 ) -> None:
+  fail = [a if not max(n, default=0) < a.ndim else no_fail
+          for n, a in zip(names, avals)]
+  if any(f is not no_fail for f in fail): raise _SpecError(fail)
+class _SpecError(Exception): pass
+
+def _check_reps(mesh, names, reps):
+  fail = [r if not _valid_repeats(mesh, r, n) else no_fail
+          for n, r in zip(names, reps)]
+  if any(f is not no_fail for f in fail): raise _RepError(fail)
+class _RepError(Exception): pass
+
+def _match_spec(mesh: Mesh, rep: Set[AxisName], dst: AxisNames, x: JaxType
+                ) -> JaxType:
+  with core.eval_context():
+    return jax.jit(HashablePartial(_match, mesh, tuple(dst.items())))(x)
+
+def _match(mesh, dst_tup, x):
+  src = P(mesh.axis_names)
+  dst = _names_to_pspec(dict(dst_tup))
+  return shard_map(_rem_singleton, mesh, (src,), dst, check_rep=False)(x)
+
+def _rem_singleton(x): return x.reshape(x.shape[1:])
+def _add_singleton(x): return x.reshape(1, *x.shape)
+
+class ShardMapTrace(core.Trace):
+  mesh: Mesh
+
+  def __init__(self, *args, mesh):
+    super().__init__(*args)
+    self.mesh = mesh
+
+  def pure(self, val):
+    val_ = _unmatch_spec(self.mesh, {}, val)
+    return ShardMapTracer(self, set(self.mesh.axis_names), val_)
+
+  def sublift(self, tracer):
+    return ShardMapTracer(self, tracer.rep, tracer.val)
+
+  def process_primitive(self, prim, tracers, params):
+    in_vals, in_rep = unzip2((t.val, t.rep) for t in tracers)
+    f = HashablePartial(_prim_applier, prim, tuple(params.items()), self.mesh)
+    with core.eval_context(), jax.disable_jit(False):
+      out_vals = jax.jit(f)(*in_vals)
+    out_rep = _rep_rules.get(prim, partial(_rep_rule, prim))(*in_rep, **params)
+    if prim.multiple_results:
+      out_rep = [out_rep] * len(out_vals) if type(out_rep) is set else out_rep
+      return map(partial(ShardMapTracer, self), out_rep, out_vals)
+    return ShardMapTracer(self, out_rep, out_vals)
+
+  def process_call(self, call_primitive, fun, tracers, params):
+    if call_primitive is not xla.xla_call_p: raise NotImplementedError
+    fun, jaxpr = _something_shady(fun)  # TODO remove when jit is initial-style
+    bind = partial(call_primitive.bind, fun)  # TODO caching (compat w/ jaxpr())
+    fake_primitive = pxla._FakePrimitive(multiple_results=True, bind=bind)
+    _rep_rules[fake_primitive] = lambda *_, **__: set()
+    out_tracers_ = self.process_primitive(fake_primitive, tracers, params)
+    out_vals = [t.val for t in out_tracers_]
+    out_rep = _output_rep(self.mesh, jaxpr(), [t.rep for t in tracers])
+    return map(partial(ShardMapTracer, self), out_rep, out_vals)
+
+@lu.transformation_with_aux
+def _something_shady(*args):
```

Comment:
Please please name it better?

Reply:
You mean with like another underscore or two on the end? 😁 

Okay, done!